File: Java_Completo_Programacao_Orientada_a_Objetos_Nelio Alves/Secao_14_Heranca_e_Polimorfismo/Override_e_Super/src/entities/SavingsAccount.java
package entities;

public class SavingsAccount extends Account{

	private Double interestRate;
	
	public SavingsAccount() {
		super();
	}

	public SavingsAccount(Integer number, String holder, Double balance, Double interestRate) {
		super(number, holder, balance);
		this.interestRate = interestRate;
	}

	public Double getInterestRate() {
		return interestRate;
	}

	public void setInterestRate(Double interestRate) {
		this.interestRate = interestRate;
	}
	
	public void updateBalance() {
		balance += balance + interestRate;
	}
	
	//Na classe Account tem a operação normal que desconta 5.0 do saque.
	//E para sobrescrever essa operação sem descontar os 5.0, é colocado o Override, ele também indica se tem algo errado na sintaxe.
	//Isso é uma sobreposição de método.
	@Override
	public void withdraw(double amount) {
		balance -= amount;
	}
	
}

File: C_Sharp_completo_programacao_orientada_a_objetos/Secao-10-Aula-138-Classes-e-metodos-selados/Secao-10-Aula-138-Classes-e-metodos-selados/Entities/SavingsAccount.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Secao_10_Aula_138_Classes_e_metodos_selados.Entities
{
    // : (significa estende), estende quem ? a classe Account.
    // sealed, não deixa criar subclasse a partir dessa classe.
    sealed class SavingsAccount : Account
    {
        public double InterestRate { get; set; }

        public SavingsAccount()
        {
        }

        // : base(), dessa forma se chama os atributos da classe que está sendo herdada, base() é a referência para a superclasse.
        public SavingsAccount(int number, string holder, double balance, double interestRate) : base(number, holder, balance)
        {
            InterestRate = interestRate;
        }

        //Método para atualizar a taxa de juros.
        public void UpdateBalance()
        {
            Balance += Balance * InterestRate;
        }

        //override, está sobreescrevendo o método da superclasse Account,
        //e esse método por estar sendo sobreescrito pode ter novas regras.
        //a regra desse método é que vai descontar a taxa de 5.0 do método que está na classe Account,
        //e depois descontar mais 2.0

        //sealed, significa que esse método não pode ser reescrito novamente em uma outra subclasse.
        //é sempre bom selar métodos sobrepostos, sobreposições multiplas pode ser uma porta de entrada para inconsistências.
        public sealed override void Withdraw(double amount)
        {
            base.Withdraw(amount);
            Balance -= 2.0;
        }
    }
}